nomad/structs/structs.go

```diff
@@ -11322,16 +11325,57 @@ func NewIdentityClaims(job *Job, alloc *Allocation, wihandle *WIHandle, wid *Wor
 		claims.JobID = job.ParentID
 	}
 
+	var taskName string
+
 	switch wihandle.WorkloadType {
 	case WorkloadTypeService:
-		claims.ServiceName = wihandle.WorkloadIdentifier
+		serviceName := wihandle.WorkloadIdentifier
+		claims.ServiceName = serviceName
+
+		// Find task name if this is a task service.
+		for _, t := range tg.Tasks {
+			for _, s := range t.Services {
+				if s.Name == serviceName {
+					taskName = t.Name
+					break
+				}
+			}
+			if taskName != "" {
+				break
+			}
+		}
+
 	case WorkloadTypeTask:
-		claims.TaskName = wihandle.WorkloadIdentifier
+		taskName = wihandle.WorkloadIdentifier
+		claims.TaskName = taskName
+
 	default:
 		// in case of an unknown workload type we quit
 		return nil
 	}
 
+	// Add ConsulNamespace and VaultNamespace claims if necessary.
+	if taskName != "" {
+		task := tg.LookupTask(taskName)
+		if task == nil {
+			return nil
+		}
+
+		if wid.IsConsul() {
+			if task.Consul != nil {
+				claims.ConsulNamespace = task.Consul.Namespace
+			} else if tg.Consul != nil {
+				claims.ConsulNamespace = tg.Consul.Namespace
+			}
```

Comment:
Unfortunately we can't because this function is also called from the scheduler, where `alloc.Job` is `nil`.
https://github.com/hashicorp/nomad/blob/573c3103fb6da48143982cd5457f9bc0a1e33221/nomad/structs/structs.go#L11293-L11298

We could have a another set of helpers at the `TaskGroup` level that is called here and also in the `Allocation.ServiceProviderNamespace*`.

Reply:
Gotcha, no worries. We can come back for that kind of tech debt cleanup later if we want.

Reply:
Yeah, it would be really nice to have a clean-up of Consul namespace handling once JWT becomes the standard way and we can have a single source of truth of how to find it at different levels.